pkg/server/handler.go

```diff
@@ -478,8 +479,81 @@ func (al *activeListener) OnNewConnection(ctx context.Context, conn types.Connec
 	conn.Start(ctx)
 }
 
+type activeProxy map[*activeConnection][]proxy.Proxy
+
+// active proxy filters,
+// holds all the information about the connection
+var filters = make(activeProxy)
+
+func (p *activeProxy) activeProxy(activeConn *activeConnection) (proxy []proxy.Proxy) {
+	proxy, _ = filters[activeConn]
+	return proxy
+}
+
+func (p *activeProxy) appendProxy(activeConn *activeConnection, proxy proxy.Proxy) []proxy.Proxy {
+	filters[activeConn] = append(filters[activeConn], proxy)
+	return filters[activeConn]
+}
+
+func (al *activeListener) activeStreamSize() int {
+	if al.conns == nil || al.conns.Len() <= 0 {
+		return 0
+	}
+	var activeStream int
+	for conn := al.conns.Front(); conn != nil; conn = conn.Next() {
+		if activeConn, ok := conn.Value.(*activeConnection); ok {
+			proxyFilters := filters.activeProxy(activeConn)
+			if proxyFilters == nil {
+				for _, rf := range activeConn.conn.FilterManager().ListReadFilter() {
+					if proxy, isProxy := rf.(proxy.Proxy); isProxy {
+						proxyFilters = filters.appendProxy(activeConn, proxy)
+					}
+				}
+			}
+			if proxyFilters != nil {
+				for _, proxyFilter := range proxyFilters {
+					activeStream += proxyFilter.ActiveStreamSize()
+				}
+			}
+		}
+	}
+
+	return activeStream
+}
+
 func (al *activeListener) OnClose() {}
 
+func (al *activeListener) PreStopHook(ctx context.Context) func() error {
```

Comment:
// 替换成以下逻辑 in future
``` go
func (al *activeListener) PreStopHook(ctx context.Context) func() error {
	// before allowing you to stop listening,
	// check that the preconditions are met.
	// for example: whether all request queues are processed ?
	return func() error {
		var remainStream int
		var waitedMilliseconds int64
		if ctx != nil {
			shutdownTimeout := ctx.Value(types.GlobalShutdownTimeout)
			if shutdownTimeout != nil {
				if timeout, err := strconv.ParseInt(shutdownTimeout.(string), 10, 64); err == nil {
					current := time.Now()
					// if there any stream being processed and without timeout,
					// we try to wait for processing to complete, or wait for a timeout.
					remainStream, waitedMilliseconds =
						al.activeStreamSize(), Milliseconds(time.Since(current))
					for ; remainStream > 0 && waitedMilliseconds <= timeout;
					remainStream, waitedMilliseconds =
						al.activeStreamSize(), Milliseconds(time.Since(current)) {
						// waiting for 10ms
						time.Sleep(10 * time.Millisecond)
						if log.DefaultLogger.GetLogLevel() >= log.DEBUG {
							log.DefaultLogger.Debugf("[activeListener] listener %s invoking stop hook, remaining stream count %d, waited time %dms",
								al.listener.Name(), remainStream, waitedMilliseconds)
						}
					}
				}
			}
		}

		if log.DefaultLogger.GetLogLevel() >= log.INFO {
			log.DefaultLogger.Infof("[activeListener] listener %s pre stop hook complete, remaining stream count %d, waited time %dms",
				al.listener.Name(), remainStream, waitedMilliseconds)
		}

		return nil
	}
}
```